Package version of What's New Guide component

donlair opened the pull request:
This PR is a refactor of the What's New component discussed in #49957.

The two big changes are:
* Moved to /packages to make the component available in both Calypso and Editing Toolkit.
* Uses `wpcom.req()` for the api fetch rather than Redux

I've also wrapped the data in the Guide component since that's how it will be used.

#### Testing instructions
* Sandbox `public-api`
* Patch your WPcom sandbox with D56831-code 
* Patch your MC sandbox with D57620-code
* Navigate to `/mobile-announce/whats-new.php` on your MC sandbox
* Add some new announcements to the tool
* Checkout this PR and navigate to `/marketing/tools/SITESLUG`
* Confirm that the What's New Guide component appears and works as expected
* Switch your language preference and confirm that API sends the correct strings
* Delete the What's New announcements that you added.

Note: It's on the marketing page right now for the sake of testing; that's not the final implementation.

Related to #
#49957 
#48722 
D56831-code
D57620-code

donlair committed:
updates

donlair committed:
updating WhatsNewGuide component

sarayourfriend commented on packages/components/src/whats-new/index.js:
```diff
@@ -0,0 +1,102 @@
+import React, { useEffect, useState } from 'react';
+import { recordTracksEvent } from '@automattic/calypso-analytics';
+import { Guide, GuidePage } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
```
We should use `react-i18n` here with the `useI18n` hook if we can.

donlair replied:
> If at all possible it would be nice to go ahead and add this in TypeScript rather than adding another regular JS file to be migrated later on. However, I know that this project has already gone through so many different iterations, so I understand if there isn't time for that. If you decide to do it and run into any trouble feel free to drop a ping for help.

@sarayourfriend do you have any objections to me going with JS for this iteration? I haven't learned TypeScript yet, but it's on my near-term todo list. I'll commit to migrating it while I go through a TS course.

sarayourfriend replied:
Sounds good to me! If you need help, there's a #typescript channel in Slack full of very friendly folks eager to help 😀

sarayourfriend reviewed (commented):
If at all possible it would be nice to go ahead and add this in TypeScript rather than adding another regular JS file to be migrated later on. However, I know that this project has already gone through so many different iterations, so I understand if there isn't time for that. If you decide to do it and run into any trouble feel free to drop a ping for help.

tyxla commented on packages/components/src/whats-new/index.js:
```diff
@@ -0,0 +1,102 @@
+			{ isLoaded && showGuide && (
+				<Guide
+					className="whats-new"
+					contentLabel={ __( "What's New at WordPress.com", 'full-site-editing' ) }
```
The textdomain (second arg) in these `__()` calls seems off. Should we remove it completely?

tyxla commented on packages/components/src/index.js:
```diff
@@ -8,3 +8,4 @@ export { default as Ribbon } from './ribbon';
 export { default as RootChild } from './root-child';
 export { default as ScreenReaderText } from './screen-reader-text';
 export { default as Suggestions } from './suggestions';
+export { default as WhatsNewGuide } from './whats-new';
```
🤔 

I expected to have this in its own package, to be honest. The components in `@automattic/components` are usually simple React components used for composing the UI of Calypso. They don't connect to Redux, nor they perform HTTP requests, record analytics / tracks events, or trigger any other side effects.

tyxla commented on packages/components/src/whats-new/index.js:
```diff
@@ -0,0 +1,102 @@
+import './style.scss';
+
+const WhatsNewGuide = () => {
+	const [ isLoaded, setIsLoaded ] = useState( false );
```
Do we need the `isLoaded` at all? Can't we just use `whatsNewData` and set it to `null` by default or something? Then if it's an array, continue with rendering below?

tyxla commented on packages/components/src/whats-new/index.js:
```diff
@@ -0,0 +1,102 @@
+					finishButtonText={ __( 'Close', 'full-site-editing' ) }
+					onFinish={ toggleWhatsNew }
+				>
+					{ whatsNewData.announcements.map( ( page, index ) => (
```
If we only need the `announcements` we might as well pass only that to `setWhatsNewData`? Or are we planning to add more data trees eventually?

tyxla commented on packages/components/src/whats-new/index.js:
```diff
@@ -0,0 +1,102 @@
+			slide_number: pageNumber,
+			is_last_slide: isLastPage,
+		} );
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```
We might as well fix that?

tyxla commented on packages/components/src/whats-new/index.js:
```diff
@@ -0,0 +1,102 @@
+					{ link && (
+						<p>
+							<a href={ link } target="_blank" rel="noreferrer">
+								Learn more
```
Does this need to be localized too?

tyxla commented on packages/components/src/whats-new/index.js:
```diff
@@ -0,0 +1,102 @@
+	);
+};
+
+function WhatsNewPage( {
```
Should we move each component to a separate file?

tyxla commented on packages/components/src/whats-new/index.js:
```diff
@@ -0,0 +1,102 @@
+			<div className="whats-new__text">
+				<h1 className="whats-new__heading">{ heading }</h1>
+				<div className="whats-new__description">
+					<p>{ description }</p>
```
Will we always have a description? Should we have it conditionally displayed only if `description` isn't falsy?

tyxla reviewed (commented):
Thanks for going the package way 👍  Nice work here, too.

I've left some thoughts, observations, and suggestions.

donlair committed:
Moving WhatsNew component to its own package.

donlair committed:
Updating component

donlair committed:
Linter fixes

tyxla commented on packages/whats-new/src/index.js:
```diff
@@ -0,0 +1,60 @@
+			Locale is: { locale }
+			{ whatsNewData && showGuide && (
+				<Guide
+					className="whats-new"
```
We'll need to change this one to `whats-new__main` or something to conform to our CSS class naming rules.

tyxla commented on packages/whats-new/src/index.js:
```diff
@@ -0,0 +1,60 @@
+			.then( ( returnedList ) => {
+				setWhatsNewData( returnedList );
+			} );
+	}, [] );
```
Seems like we depend on `locale` here so we should include it in the list of dependencies. Because it's expected that we trigger a new request when the locale changes, right?

tyxla commented on packages/whats-new/package.json:
```diff
@@ -0,0 +1,56 @@
+	},
+	"peerDependencies": {
+		"react": "^16.8",
+		"react-dom": "^16.8"
```
🤔  Any idea what we need `react-dom` for?

tyxla commented on packages/whats-new/src/index.js:
```diff
@@ -0,0 +1,60 @@
+							key={ page.announcementId }
+							pageNumber={ index + 1 }
+							isLastPage={ index === whatsNewData.length - 1 }
+							{ ...page }
```
Any chance `page` will contain more than we'll ever need? At a first glance, I can see it contains a `announcementId` property, but we don't need that anywhere. Should we more specifically list what we expect `page` to contain and pass further down?

tyxla commented on packages/whats-new/src/index.js:
```diff
@@ -0,0 +1,60 @@
+			{ whatsNewData && showGuide && (
+				<Guide
+					className="whats-new"
+					contentLabel={ __( "What's New at WordPress.com" ) }
```
A note on introducing new strings: we should be careful about what strings are final and what aren't. Usually, when we're not sure if a string is final, we'll not pass it through the localization functions until we've confirmed that.

tyxla commented on packages/whats-new/src/index.js:
```diff
@@ -0,0 +1,60 @@
+	return (
+		<>
+			Locale is: { locale }
+			{ whatsNewData && showGuide && (
```
Do you think we could move those checks to return `null` earlier above than here? It would reduce the indentation level (therefore, logic depth) a bit.

tyxla commented on packages/whats-new/src/whats-new-page.js:
```diff
@@ -0,0 +1,55 @@
+	const __ = useI18n().__;
+
+	useEffect( () => {
+		recordTracksEvent( 'calypso_block_editor_whats_new_slide_view', {
```
FWIW event names are expected to be of the form: `<source>_<context>_<optional subcontext>_<action>_<optional qualifier>` - source: PCYsg-4S2-p2

tyxla commented on packages/whats-new/src/whats-new-page.js:
```diff
@@ -0,0 +1,55 @@
+	return (
+		<GuidePage className="whats-new-page__container">
+			<div className="whats-new-page__text">
+				<h1 className="whats-new-page__heading">{ heading }</h1>
```
Any chance heading might not exist? Should we check before we render the `<h1>`?

tyxla commented on packages/whats-new/src/whats-new-page.js:
```diff
@@ -0,0 +1,55 @@
+			<div className="whats-new-page__text">
+				<h1 className="whats-new-page__heading">{ heading }</h1>
+				<div className="whats-new-page__description">
+					<p>{ description }</p>
```
Any chance description might not exist? Should we check before we render the `<p>`?

tyxla commented on packages/whats-new/src/whats-new-page.js:
```diff
@@ -0,0 +1,55 @@
+					<p>{ description }</p>
+					{ link && (
+						<p>
+							<a href={ link } target="_blank" rel="noreferrer">
```
I'm assuming you're still working on that part, but this appears like a single link within a paragraph. Were you planning to make it a button or something? Or at least add a more specific `className` to it, so we can more easily style it in the future?

tyxla commented on packages/whats-new/src/whats-new-page.js:
```diff
@@ -0,0 +1,55 @@
+					) }
+				</div>
+			</div>
+			<div className="whats-new-page__visual">
```
Any chance the image may not exist and we would output an empty placeholder?

tyxla commented on packages/whats-new/src/style.scss:
```diff
@@ -0,0 +1,197 @@
+.whats-new-page__visual {
+	height: 40%;
+	padding: 25px 25px 0;
+	background: #1581d8;
```
Using colors like this isn't recommended and won't guarantee that we consider the user's current color scheme. Could we use variables for all the colors instead?

tyxla commented on packages/whats-new/package.json:
```diff
@@ -0,0 +1,56 @@
+	"repository": {
+		"type": "git",
+		"url": "git+https://github.com/Automattic/wp-calypso.git",
+		"directory": "packages/your-package"
```
```suggestion
		"directory": "packages/whats-new"
```

tyxla commented on packages/whats-new/package.json:
```diff
@@ -0,0 +1,56 @@
+		"wordpress"
+	],
+	"author": "Automattic Inc.",
+	"contributors": [],
```
Seems like we can rather ditch this one if it's going to be empty.

tyxla commented on packages/whats-new/src/whats-new-page.js:
```diff
@@ -0,0 +1,55 @@
+					src={ imageSrc }
+					alt={ description }
+					aria-hidden="true"
+					className={ 'whats-new-page__image' + ( alignBottom ? ' align-bottom' : '' ) }
```
Could we use the `classNames` library to construct those? Would it add much overhead?

donlair replied:
Removed that conditional class since I ended up not using it anyways.

tyxla commented on packages/whats-new/src/index.js:
```diff
@@ -0,0 +1,60 @@
+import React, { useEffect, useState } from 'react';
+import { Guide } from '@wordpress/components';
+import { useI18n } from '@automattic/react-i18n';
+import { useLocale } from '@automattic/i18n-utils';
```
We'll need to include this one in the list of dependencies in package.json.

tyxla commented on packages/whats-new/package.json:
```diff
@@ -0,0 +1,56 @@
+{
```
Seems like this will introduce changes to `yarn.lock`. Let's make sure to add them to the PR.

tyxla reviewed (commented):
This looks really great, solid start @donlair 🙌 

We should make all checks green. Part of why it's failing is because there are uncommited changes to `yarn.lock`.

I've also left some comments and suggestions. Let me know what you think 😉

donlair committed:
rebase

donlair committed:
Working on addressing feedback

donlair committed:
style updates

donlair committed:
style tweaks

sarayourfriend commented on packages/whats-new/package.json:
```diff
@@ -40,17 +39,16 @@
 		"@automattic/calypso-build": "^7.0.0",
-		"react": "^16.12.0",
-		"react-dom": "^16.12.0"
+		"react": "^16.12.0"
```
Why is `react` duplicated in peer and dev dependencies? Is this just a pattern I'm not familiar with?

donlair replied:
Nope, just me not really understanding how the dependencies work :)

sarayourfriend commented on packages/whats-new/src/style.scss:
```diff
@@ -131,7 +132,8 @@ $wpcom-modal-footer-height: 30px + ( $wpcom-modal-footer-padding-v * 2 );
 	padding: 25px 25px 0;
-	background: #1581d8;
+	// background: #1581d8;
+	background: var( --studio-blue );
```
Is `--studio-blue` the correct theme variable to use or is there a different like `--color-primary`?

sarayourfriend commented on packages/whats-new/src/whats-new-page.js:
```diff
@@ -27,9 +28,9 @@ function WhatsNewPage( {
+					{ description && <p>{ description }</p> }
 					{ link && (
 						<p>
 							<a href={ link } target="_blank" rel="noreferrer">
```
Nit: Can we use `@wordpress/components` `Button` here?

sarayourfriend commented on packages/whats-new/src/whats-new-page.js:
```diff
@@ -27,9 +28,9 @@ function WhatsNewPage( {
 		<GuidePage className="whats-new-page__container">
 			<div className="whats-new-page__text">
-				<h1 className="whats-new-page__heading">{ heading }</h1>
+				{ heading && <h1 className="whats-new-page__heading">{ heading }</h1> }
```
Nit: Can this use `@wordpress/components` `Text` with variant set to `title.large`?

donlair replied:
Note: Text is still experimental so we decided against using it for now.

donlair replied:
Note: Text is still experimental so we decided to not use it for now.

donlair commented:
Thanks @sarayourfriend! I'll work through your comments now :)

donlair committed:
Addressing review feedback

donlair committed:
Package.json update

donlair commented:
Thanks for the reviews @tyxla and @sarayourfriend! I think your comments have all been addressed now!

sarayourfriend commented on packages/whats-new/src/whats-new-page.js:
```diff
@@ -28,15 +29,19 @@ function WhatsNewPage( {
+							isLink
+							target="_blank"
+						>
+							{ __( 'Learn more' ) }
```
Sorry I didn't catch this before, but do we need to pass `__i18n_text_domain__` to `__` like we do in other packages?

sarayourfriend commented on packages/whats-new/src/whats-new-page.js:
```diff
@@ -3,7 +3,8 @@
 import React, { useEffect } from 'react';
 import { recordTracksEvent } from '@automattic/calypso-analytics';
-import { GuidePage } from '@wordpress/components';
+// eslint-disable-next-line wpcalypso/no-unsafe-wp-apis
```
Sorry about this, but I didn't realize `Text` was experimental still 😢 I think we should probably not use it if that is the case. What do you think?

donlair committed:
Removing experimental Text component, adding translation context, and removing unused className check

donlair committed:
Removing component placement from Marketing Tools

donlair commented:
@sarayourfriend @tyxla I believe this changes address everything. I also removed the placeholder WhatsNew component insert in Marketing Tools and rebased.

tyxla reviewed (commented):
Things are generally looking great in this PR. Really nice work 👏 

There seem to be too many unrelated changes in yarn.lock. Is there any chance you recreated it from scratch? If that's the case, I'd suggest reverting to the original `yarn.lock` in `trunk` and running `yarn` / `yarn install` from there.

donlair commented:
> Things are generally looking great in this PR. Really nice work 👏
> 
> There seem to be too many unrelated changes in yarn.lock. Is there any chance you recreated it from scratch? If that's the case, I'd suggest reverting to the original `yarn.lock` in `trunk` and running `yarn` / `yarn install` from there.

Man, I've been fighting that file this entire PR! Let me see if I can finally fix it :)

donlair committed:
attempting to fix yarn.lock

donlair committed:
reverting yarn.lock

donlair committed:
Reverted yarn.lock and then ran yarn && yarn start

sarayourfriend commented on packages/whats-new/src/whats-new-page.js:
```diff
@@ -0,0 +1,53 @@
+							isLink
+							target="_blank"
+						>
+							{ __( 'Learn more', 'Whats New Announcements' ) }
```
I think there was a misunderstanding about how to do the `__i18n_text_domain__` trick for i18n-ized packages. There's more information here: pbAok1-1Ao-p2

```suggestion
							{ __( 'Learn more', __i18n_text_domain__ ) }
```

Basically you'll want to apply the `__i18n_text_domain__ ` argument to each call to `__` like I did in this suggestion.

donlair committed:
Reverting yarn.lock

donlair committed:
Updating i18n domain

sarayourfriend commented on packages/whats-new/.eslintrc.js:
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+	rules: {
+		'react/react-in-jsx-scope': 0,
+	},
```
This shouldn't be necessary, is it?

donlair replied:
nope, there were a couple changes that I hadn't saved, about to update them

donlair committed:
Remainder of i18n updates

donlair committed:
Missed one save

tyxla commented on packages/whats-new/package.json:
```diff
@@ -0,0 +1,51 @@
+		"wpcom": "^5.4.2"
+	},
+	"devDependencies": {
+		"wpcom-proxy-request": "^6.0.0",
```
Does this look like it should be a dependency and not a dev dependency?

tyxla commented on packages/whats-new/src/index.js:
```diff
@@ -0,0 +1,60 @@
+
+	return (
+		<Guide
+			className="src__main"
```
This classname doesn't feel right. Not only it doesn't correspond to the package / component it lives in, but if other packages do the same, they'll conflict with each other.

donlair replied:
the linter said it should be called `src__`, but you're right. I'll disable the linter for that line.

tyxla commented on packages/whats-new/src/index.js:
```diff
@@ -0,0 +1,60 @@
+	return (
+		<Guide
+			className="src__main"
+			contentLabel={ __( "What's New at WordPress.com", __i18n_text_domain__ ) }
```
Did we intentionally capitalize the words here?

tyxla commented on packages/whats-new/src/whats-new-page.js:
```diff
@@ -0,0 +1,53 @@
+			<div className="whats-new-page__visual">
+				{ imageSrc && (
+					<img
+						key={ key }
```
I'm assuming that using the same key per announcement and not per image is intentional?

donlair replied:
I went ahead and removed since I'm not using the key for anything in particular!

donlair committed:
Addressing feedback and tweaking dependencies to reduce yarn.lock changes

donlair committed:
rebase

a8ci18n commented on packages/whats-new/src/index.js:
```diff
@@ -0,0 +1,60 @@
+		<Guide
+			// eslint-disable-next-line wpcalypso/jsx-classname-namespace
+			className="whats-new-guide__main"
+			contentLabel={ __( "What's new at WordPress.com", __i18n_text_domain__ ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **24** times:
`translate( 'What's New at WordPress.com' )` *ES Score: 12*

a8ci18n replied:
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **24** times:
`translate( 'What's New at WordPress.com' )` *ES Score: 12*

donlair committed:
trying to get build tests to pass yarn.lock changes

donlair committed:
more work on yarn

a8ci18n commented on packages/whats-new/src/index.js:
```diff
@@ -0,0 +1,60 @@
+		<Guide
+			// eslint-disable-next-line wpcalypso/jsx-classname-namespace
+			className="whats-new-guide__main"
+			contentLabel={ __( "What's new at WordPress.com", __i18n_text_domain__ ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **24** times:
`translate( 'What's New at WordPress.com' )` *ES Score: 12*

tyxla replied:
Seems like we already have a translated version of this, so let's reuse it. Sorry for the back and forth! 

```suggestion
			contentLabel={ __( "What's New at WordPress.com", __i18n_text_domain__ ) }
```

donlair commented:
@tyxla I think I finally resolved those yarn.lock issues, and addressed your feedback!

tyxla reviewed (approved):
This looks great to me! Once we revert the localized string, let's 🚢  

Nice work @donlair 👏

donlair committed:
Rebase and string update

donlair committed:
Rebase
